sentry-ruby/spec/sentry/metrics/aggregator_spec.rb

```diff
@@ -0,0 +1,297 @@
+        Sentry.get_current_scope.set_transaction_name('foo')
+        subject.add(:c, 'incr', 1)
+        _, _, _, tags = subject.buckets.values.first.keys.first
+        expect(tags).to include(['transaction', 'foo'])
+      end
+
+      it 'does not has the low quality transaction name in tags serialized in the bucket metric key' do
+        Sentry.get_current_scope.set_transaction_name('foo', source: :url)
+        subject.add(:c, 'incr', 1)
+        _, _, _, tags = subject.buckets.values.first.keys.first
+        expect(tags).not_to include(['transaction', 'foo'])
+      end
+    end
+
+    it 'creates a new CounterMetric instance if not existing' do
+      expect(Sentry::Metrics::CounterMetric).to receive(:new).and_call_original
+      subject.add(:c, 'incr', 1)
+
+      metric = subject.buckets.values.first.values.first
+      expect(metric).to be_a(Sentry::Metrics::CounterMetric)
+      expect(metric.value).to eq(1.0)
+    end
+
+    it 'reuses the existing CounterMetric instance' do
+      allow(Time).to receive(:now).and_return(fake_time)
+
+      subject.add(:c, 'incr', 1)
+      metric = subject.buckets.values.first.values.first
+      expect(metric.value).to eq(1.0)
+
+      expect(Sentry::Metrics::CounterMetric).not_to receive(:new)
+      expect(metric).to receive(:add).with(2).and_call_original
+      subject.add(:c, 'incr', 2)
+      expect(metric.value).to eq(3.0)
+    end
+
+    it 'creates a new DistributionMetric instance if not existing' do
+      expect(Sentry::Metrics::DistributionMetric).to receive(:new).and_call_original
+      subject.add(:d, 'dist', 1)
+
+      metric = subject.buckets.values.first.values.first
+      expect(metric).to be_a(Sentry::Metrics::DistributionMetric)
+      expect(metric.value).to eq([1.0])
+    end
+
+    it 'creates a new GaugeMetric instance if not existing' do
+      expect(Sentry::Metrics::GaugeMetric).to receive(:new).and_call_original
+      subject.add(:g, 'gauge', 1)
+
+      metric = subject.buckets.values.first.values.first
+      expect(metric).to be_a(Sentry::Metrics::GaugeMetric)
+      expect(metric.serialize).to eq([1.0, 1.0, 1.0, 1.0, 1])
+    end
+
+    it 'creates a new SetMetric instance if not existing' do
+      expect(Sentry::Metrics::SetMetric).to receive(:new).and_call_original
+      subject.add(:s, 'set', 1)
+
+      metric = subject.buckets.values.first.values.first
+      expect(metric).to be_a(Sentry::Metrics::SetMetric)
+      expect(metric.value).to eq(Set[1])
+    end
+  end
+
+  describe '#flush' do
+    context 'with empty buckets' do
+      it 'returns early and does nothing' do
+        expect(sentry_envelopes.count).to eq(0)
+        subject.flush
+      end
+
+      it 'returns early and does nothing with force' do
+        expect(sentry_envelopes.count).to eq(0)
+        subject.flush(force: true)
+      end
+    end
+
+    context 'with pending buckets' do
+      before do
+        allow(Time).to receive(:now).and_return(fake_time)
```

Comment:
Maybe we should start using [timecop](https://github.com/travisjeffery/timecop) for these tests.

Reply:
for now it's simple enough, if you really want I can refactor all the time related tests separately.